plugins/modules/opensearch.py

```diff
@@ -0,0 +1,1194 @@
+        return None
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Couldn't get domain {0}".format(domain_name))
+    return response["DomainStatus"]
+
+
+def get_domain_config(client, module, domain_name):
+    try:
+        response = client.describe_domain_config(DomainName=domain_name)
+    except client.exceptions.ResourceNotFoundException:
+        return None
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Couldn't get domain {0}".format(domain_name))
+    return response["DomainConfig"]
+
+
+def domain_to_facts(client, module, domain):
+    try:
+        domain["Tags"] = boto3_tag_list_to_ansible_dict(
+            client.list_tags(ARN=domain["ARN"], aws_retry=True)["TagList"]
+        )
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(
+            e, "Couldn't get tags for domain %s" % domain["domain_name"]
+        )
+    except KeyError:
+        module.fail_json(msg=str(domain))
+
+    return camel_dict_to_snake_dict(domain, ignore_list=["Tags"])
+
+
+def wait_for_domain_status(client, module, domain_name, waiter_name):
+    if not module.params["wait"]:
+        return
+    timeout = module.params["wait_timeout"]
+    deadline = time.time() + timeout
+    while time.time() < deadline:
+        status = get_domain(client, module, domain_name)
+        if status is None:
+            if waiter_name == "domain_deleted":
+                return
+        else:
+            if (
+                waiter_name == "domain_available"
+                and status["Created"]
+                and not status["Processing"]
+                and not status["UpgradeProcessing"]
+            ):
+                return
+        time.sleep(15)
+
+
+def ensure_domain_absent(client, module):
+    domain_name = module.params.get("domain_name")
+    changed = False
+
+    domain = get_domain(client, module, domain_name)
+    try:
+        client.delete_domain(DomainName=domain_name)
+        changed = True
+    except (
+        botocore.exceptions.ClientError,
+        botocore.exceptions.BotoCoreError,
+    ) as e:
+        module.fail_json_aws(e, msg="trying to delete domain")
+
+    # If we're not waiting for a delete to complete then we're all done
+    # so just return
+    if not domain or not module.params.get("wait"):
+        return dict(changed=changed)
+    try:
+        wait_for_domain_status(client, module, domain_name, "domain_deleted")
+        return dict(changed=changed)
+    except client.exceptions.ResourceNotFoundException:
+        return dict(changed=changed)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, "awaiting domain deletion")
+
+
+def ensure_tags(client, module, resource_arn, existing_tags, tags, purge_tags):
```

Comment:
I would personally move all the util methods inside a dedicated file here  https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils. 

@markuman what do you think about? Does this make sense for you?

Reply:
Is this a new pattern? I see only two modules externalize functions under module_utils. It seems this would make it harder to locally set or override the `opensearch` module. I.e. when a module has a bug or needs some enhancement, if it's self contained users can just copy the `opensearch.py` file and configure the library search path. If the functionality is split, then it makes it a bit harder.

Reply:
Many modules import module_utils hosted in amazon.aws. I suggested to do so because if we add another module (for example opensearch_info.py), we can directly import the needed functions from module_utils instead of having to rewrite them.
The main objective of module_utils is to host helpers that can be used multiple times by different modules. in this way, we write the code once and then reduce redundancy and errors.

By the way, for the split, I would wait for @markuman's confirmation.

Reply:
Are you suggesting to import from `amazon.aws` or from `community.aws`? You are referencing both. In `community.aws`, there are only two files. I'm not too keen on adding these files to `amazon.aws`, this is most likely going to make the PR process much more involved, right? Open 2 PRs, make sure both are approved, make sure they work together, and if subsequently a change needs to be made, refactors would be much harder to do.

Reply:
I suggested to move the helper methods in a new opensearch.py file here  https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils. Then, you can import them in your module (no need to open the PR in amazon.aws).

@tremble  What do you think about moving `domain to_facts()` -  maybe we can rename this method to get_tags for example, `ensure_tags()` and `get_domain()` inside https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils?

Reply:
@sebastien-rosset

Unfortunately, we don't have a lot of historical consistency in the way our modules have generally been written, so I ask for your understanding when it comes to "but other modules don't do that".  Cleaning up is a massive job, and is getting picked at very slowly.

Helper functions that can be used by other services should live in `amazon.aws`.  However, anything OpenSearch specific should live in `community.aws` unless or until the modules themselves move to `amazon.aws`.  Personally I'd stick to only making changes here in `community.aws` for now, we can talk about refactoring things over into amazon.aws later if appropriate.  The kinds of OpenSearch specific helpers @alinabuzachis's talking about would move into `plugins/module_utils/opensearch.py`.

FWIW, some modules ended up being split across amazon.aws and community.aws initially and it's been more painful to test and manage than was predicted (there was a reasonable logic behind it, but "hindsight").

You don't see many modules using module_utils today because things have grown organically and it was "complicated" prior to the Ansible- Core/Collections split.  However, the some_service/some_service_info model has resulted in far more copy and paste than it should because people don't use module_utils.  Starting cleanly by moving generic code into module_utils/ is technically the better idea.

Specifically with regards "XYZ_to_facts" as a name, we almost never return **facts** from our modules these days.  I tend to refer to is as "normalize_XYZ", but it's a good example of a piece of code best handled in module_utils, because it'll be needed/wanted by both opensearch and a potential future opensearch_info module.

Reply:
Sure, no problem I understand the challenges of history. If I go on the path of splitting the module to externalize functions to `module_utils`, I need to find a way to use the `opensearch` module in a environment until `community.aws` version 3.1 is released. Reviewing PRs, releasing software and testing the new software version usually takes weeks or months, meanwhile I would like to be able to use the `opensearch` module for our needs.

When the module is self contained as a single file, the transition is very easy: modify `ansible.cfg`, add a `library = ./library` statement, drop a copy of `opensearch.py` under `./library`, and that's it. We usually commit the new or patched module to our repos. Then later when changes have been upstreamed, released and tested, we can upgrade the version of the ansible package and remove the locally patched module.

When the module is not self contained, the transition period looks more complicated. I hope that does not mean we have to build the `community.aws` fork and install it in every developer environment as a replacement of the upstream version.

Reply:
Is this what you had in mind? b668e01ad381b54c7299b395e44c1674020005b6

Reply:
@sebastien-rosset yes, exactly! You should also move `ensure_tags()` to module_utils. Thanks.

Reply:
done